[extractor/twitch] Skip errors in GQL response

elyse0 opened the pull request:
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

<!--

Explanation of your *pull request* in arbitrary form goes here. Please **make sure the description explains the purpose and effect** of your *pull request* and is worded well enough to be understood. Provide as much **context and examples** as possible

-->

This PR fixes the problem where Twitch's GQL API sometimes returns an error before the actual data, breaking the extractor that assumed the video data would be at index 0.

Full json response: https://gist.github.com/elyse0/3b1679b6a0b8d7aa8112f75c1eb52403

#### Trimmed json
```json
[
  {
    "errors": [
      {
        "message": "PersistedQueryNotFound"
      }
    ],
    "extensions": {
      "durationMilliseconds": 3,
      "operationName": "VideoMetadata",
      "requestID": "01GSTXCZDDA9GN7ZHBRSB9MP4Y"
    }
  },
  {
    "data": {
      "video": {
        "id": "1697007802",
        "moments": {
          ...
          "__typename": "VideoMomentConnection"
        },
        "__typename": "Video"
      }
    },
    "extensions": {
      "durationMilliseconds": 61,
      "operationName": "VideoPlayer_ChapterSelectButtonVideo",
      "requestID": "01GSTXCZDDA9GN7ZHBRSB9MP4Y"
    }
  },
  {
    "data": {
      "video": {
        "id": "1697007802",
        "seekPreviewsURL": "https://d1m7jfoe9zdc1j.cloudfront.net/66e41701aa009d198649_vickypalami_40717337225_1672796209/storyboards/1697007802-info.json",
        "__typename": "Video"
      }
    },
    "extensions": {
      "durationMilliseconds": 25,
      "operationName": "VideoPlayer_VODSeekbarPreviewVideo",
      "requestID": "01GSTXCZDDA9GN7ZHBRSB9MP4Y"
    }
  }
]

```

Fixes #6308


<details open><summary>Template</summary> <!-- OPEN is intentional -->

<!--

# PLEASE FOLLOW THE GUIDE BELOW

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes `[ ]` relevant to your *pull request* (like [x])
- Use *Preview* tab to see how your *pull request* will actually look like

-->

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

pukkandan commented on yt_dlp/extractor/twitch.py:
```diff
@@ -380,13 +380,16 @@ def _download_info(self, item_id):
+        initial_index = 1 if data[0].get('errors') else 0
 
+        video = traverse_obj(data, (initial_index, 'data', 'video'))
         if video is None:
-            raise ExtractorError(
-                'Video %s does not exist' % item_id, expected=True)
+            raise ExtractorError(f'Video {item_id} does not exist', expected=True)
+
+        video['moments'] = traverse_obj(data, (initial_index + 1, 'data', 'video', 'moments', 'edges', ..., 'node'))
+        video['storyboard'] = traverse_obj(
+            data, (initial_index + 2, 'data', 'video', 'seekPreviewsURL'), expected_type=url_or_none)
```
```suggestion
        video = traverse_obj(data, (..., 'data', 'video'))
        if not video:
            raise ExtractorError(f'Video {item_id} does not exist', expected=True)

        video['moments'] = traverse_obj(data, (..., 'data', 'video', 'moments', 'edges', ..., 'node'))
        video['storyboard'] = traverse_obj(data, (..., 'data', 'video', 'seekPreviewsURL',{url_or_none}), get_all=False)

```
Wouldn't this work?

elyse0 replied:
Mmm it doesn't

The bug is quite rare, but to debug this I load the json I posted above

```python
f = open('/home/amish/Documents/Programming/yt-dlp/tmp/twitch-error.json')
data = json.load(f)

video = traverse_obj(data, (..., 'data', 'video'))
...
```

```bash
[debug] Command-line config: ['https://www.twitch.tv/videos/1536751224', '--verbose', '-F']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2023.02.17 [a0a7c0154] (source)
[debug] Lazy loading extractors is disabled
[debug] Git HEAD: 5d75e21fa
[debug] Python 3.10.7 (CPython x86_64 64bit) - Linux-5.19.0-31-generic-x86_64-with-glibc2.36 (OpenSSL 3.0.5 5 Jul 2022, glibc 2.36)
[debug] exe versions: ffmpeg N-108931-g4dda3b1653-20221104 (setts), ffprobe N-108931-g4dda3b1653-20221104, phantomjs broken, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.15.0, certifi-2022.09.24, mutagen-1.46.0, sqlite3-2.6.0, websockets-10.4
[debug] Proxy map: {}
[debug] Loaded 1782 extractors
[twitch:vod] Extracting URL: https://www.twitch.tv/videos/1536751224
[twitch:vod] 1536751224: Downloading stream metadata GraphQL
ERROR: list indices must be integers or slices, not str
Traceback (most recent call last):
  File "/home/amish/Documents/Programming/yt-dlp/yt_dlp/YoutubeDL.py", line 1502, in wrapper
    return func(self, *args, **kwargs)
  File "/home/amish/Documents/Programming/yt-dlp/yt_dlp/YoutubeDL.py", line 1578, in __extract_info
    ie_result = ie.extract(url)
  File "/home/amish/Documents/Programming/yt-dlp/yt_dlp/extractor/common.py", line 693, in extract
    ie_result = self._real_extract(url)
  File "/home/amish/Documents/Programming/yt-dlp/yt_dlp/extractor/twitch.py", line 520, in _real_extract
    video = self._download_info(vod_id)
  File "/home/amish/Documents/Programming/yt-dlp/yt_dlp/extractor/twitch.py", line 390, in _download_info
    video['moments'] = traverse_obj(data, (..., 'data', 'video', 'moments', 'edges', ..., 'node'))
TypeError: list indices must be integers or slices, not str
```

pukkandan replied:
mb. It needs get_all=False
```diff
-        video = traverse_obj(data, (..., 'data', 'video'))
+        video = traverse_obj(data, (..., 'data', 'video'), get_all=False)
```
Edited above comment accordingly

pukkandan replied:
`...` with `get_all=False` looks at all indices, returning the first one. So unless the json can contain unrelated videos, this would be better than checking just 0/1 index

garoto commented:
With this patch applied, stream titles will randomly fail to be fetched. It changes on every run of the `for loop` below:

```
> for /f %a in (vid) do @yt-dlp --get-title %a
Untitled Broadcast
The game in the year of our lord 2023 without FOV options 🤣🤣🤣
BIRTHDAY SPEEDRUNS
Returnal
Untitled Broadcast
SPOOKY SUNDAY - The Karaoke
Untitled Broadcast
Any Goblin Caves Enjoyers?

[ R:\ ]
> for /f %a in (vid2) do @yt-dlp --get-title %a
Untitled Broadcast
Untitled Broadcast
Untitled Broadcast
Elden Ring Any% Glitchless WR Attempts | Atomic Hearts/Rivals Later w/ !team
Untitled Broadcast
Untitled Broadcast
PURE DARKNESS Elden Ring in FIRST PERSON | Horror Mode | !Vods (I won the !KemonoQuest)
Beating Elden Ring in FIRST PERSON | All Remembrances | !Vods (I won the !KemonoQuest)
```

`vid` contents: https://0x0.st/Hs3g.txt
`vid2` contents: https://0x0.st/Hs3E.txt

My `yt-dlp` environment:

```
[debug] Command-line config: ['-v']
[debug] User config "C:\Users\garoto\yt-dlp.conf": ['--extractor-args', 'youtube:skip=translated_subs', '--external-downloader-args', '-hide_banner -loglevel warning -stats', '--no-sponsorblock', '--no-mtime', '--no-warnings', '--ignore-errors', '--force-ipv4', '--restrict-filenames', '--merge-output-format', 'mkv', '--sub-lang', 'en.*,pt,pt-BR', '--write-sub', '--embed-subs', '--convert-subs', 'srt', '--format', '(bestvideo[width<=1920][fps<=60][vcodec^=vp9]+bestaudio[acodec=opus])/299+140/137+251/best/bestvideo+bestaudio', '--output', '%(title).70s-%(upload_date)s-%(uploader)s-%(extractor_key)s_%(id).19s.%(ext)s']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8 (No VT), error utf-8 (No VT), screen utf-8 (No VT)
[debug] yt-dlp version 2023.02.17 [a0a7c01] (pip)
[debug] Python 3.8.2 (CPython AMD64 64bit) - Windows-8.1-6.3.9600-SP0 (OpenSSL 1.1.1d  10 Sep 2019)
[debug] exe versions: ffmpeg 4.0.2, ffprobe 4.0.2
[debug] Optional libraries: Cryptodome-3.17, brotli-1.0.9, certifi-2022.12.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-10.4
[debug] Proxy map: {}
[debug] Loaded 1782 extractors
```

Grub4K commented:
I think hardcoding an index of 1 might not be what we want. Can we traverse through the response using `...` instead maybe?

elyse0 commented:
Looking at the extractor and error more closely, the problem was that some _OPERATION_HASHES were not up to date

I've updated these hashes and updated some tests

This shoud fix the titles @garoto

garoto commented:
> This shoud fix the titles @garoto

Tested a bunch of VODs and it seems to be fixed indeed. Thank you!
